aten/src/ATen/native/NativeFunctions.cpp

```diff
@@ -584,5 +586,77 @@ Tensor RoiPooling2d_backward_cpu(
   throw std::runtime_error("not implemented");
 }
 
+
+
+// TODO Replace this with more accurate digamma().
+template <typename Scalar>
+static inline Scalar digamma_one(Scalar x) {
+  const Scalar eps = x * 1e-2;
+  return (std::lgamma(x + eps) - std::lgamma(x - eps)) / (eps + eps);
+}
+
+/** Computes the reparameterized gradient -(d/dalpha cdf(x;alpha)) / pdf(x;alpha)
+    for random number x drawn from a standard Gamma distribution Gamma(alpha).
+*/
+template <typename Scalar>
+static inline Scalar standard_gamma_grad_one(Scalar x, Scalar alpha) {
+  // Use an asymptotic approximation for small x.
+  if (x < 0.2f) {
+    const Scalar a0 = 1 / alpha;
+    const Scalar a1 = 1 / (alpha + 1);
+    const Scalar a2 = 1 / (alpha + 2);
+    const Scalar pow_x_alpha = std::pow(x, alpha);
+    const Scalar gamma_pdf = std::pow(x, alpha - 1) * std::exp(-x);
+    const Scalar gamma_cdf = pow_x_alpha * (a0 - x*a1 + 0.5f*x*x*a2);
+    const Scalar gamma_cdf_alpha = (std::log(x) - digamma_one(alpha)) * gamma_cdf
+        - pow_x_alpha * (a0*a0 - x*a1*a1 + 0.5f*x*x*a2*a2);
+    const Scalar result = -gamma_cdf_alpha / gamma_pdf;
+    return std::isnan(result) ? 0 : result;
+  }
+
+  // Use an asymptotic approximation for large alpha.
+  if (alpha > 50.0f) {
+    return std::sqrt(x / alpha);
+  }
+
+  // Use a bivariate rational approximation to the reparameterized gradient.
+  const Scalar u = std::log(x / alpha);
+  const Scalar v = std::log(alpha);
+  static const Scalar coef_uv[3][8] = {
+    {0.16028008, -0.088064309, 0.019630876, -0.0016920282,
+     1.0, 0.36659853, 0.10843863, 0.0066895454},
+    {0.521894, 0.16095838, 0.06237597, 0.0023884253,
+     0.083457714, 0.0073297628, -0.0059299053, -0.00093720389},
+    {-0.0031143957, -0.012143877, -0.0057656484, -0.00064847254,
+     0.0087262576, -0.00022820524, 1.8871047e-05, 9.6307964e-06},
+  };
+  Scalar coef_v[8];
+  for (int i = 0; i < 8; ++ i) {
+    coef_v[i] = coef_uv[0][i] + u * (coef_uv[1][i] + u * coef_uv[2][i]);
+  }
+  const Scalar p = coef_v[0] + v * (coef_v[1] + v * (coef_v[2] + v * coef_v[3]));
+  const Scalar q = coef_v[4] + v * (coef_v[5] + v * (coef_v[6] + v * coef_v[7]));
+  return std::exp(p / q);
+}
+
+template <typename Scalar>
+struct StandardGammaGradOp {
+  void operator()(Scalar& ret_val, const Scalar& self_val, const Scalar &alpha_val, bool& early_exit)
```

Comment:
I guess we're using non-const references here because it makes multiple-return easier?

Reply:
Although you call it out explicitly in the PR text, I am a bit perplexed why `early_exit` is needed. gamma doesn't seem to exercise this case.

Reply:
I first implemented all_close, but the GPU version really requires reduction (you can see this in the different implementations of equal in TH/THC), so I scrapped it.  But all_close would use it.

Reply:
I'm not a fan of early_exit: I don't think it will work well with parallelization strategies on the CPU. Currently, the default for apply functions is for serial execution. If we care about CPU performance, the default should be parallel, which means the API shouldn't expose serial semantics.